### ISSUE TITLE
Limitation in css-calc support

### ISSUE THREAD

**Issue by m-abs** [NativeScript/NativeScript#7702]
**Environment**
Provide version numbers for the following components (information can be retrieved by running `tns info` in your project folder or by inspecting the `package.json` of the project):
 - CLI: 6+
 - Cross-platform modules: next
 - Android Runtime:
 - iOS Runtime:
 - Plugin(s):

Feature was just merged and will be available in nativescript 6.1.x.

**Describe the bug**

After merging #7553 "feat: implement css-variables and css-calc" there is a limitation in the implementation of `css-calc` in NativeScript.

Expressions like `calc(100% - 10px)` are reduced to `calc(100% - 10px)` because we don't know what the relative size of `100%` is at the time.

**To Reproduce**

Create a CSS rule like this:
```css
element {
   width: calc(100% - 10px);
}
```

This rule will log an error from `PercentLength.parse(...)` or `Length.parse(...)` because of an invalid value.

**Expected behavior**
The `css calc` expression should work, in this case width should be 100% of the parent's width minus 10px.

**Additional context**

**Comment by m-abs** [comment]
I think the solution for this could also solve #3707

To implement this, we need to a way to get the relative size for the `100%` per property (width, height, padding, margin, translateX/Y).
This could be done by adding a new function to the options for `CssProperty` and `CssAnimationProperty` to get the size value and extending `PercentLength.parse` to handle these `calc(...)`-expressions.

**Comment by vallemar** [comment]
is anything known about this?

**Comment by m-abs** [comment]
@vallemar Unfortunately no.

**Comment by vallemar** [comment]
@rigor789 I am trying to close this problem as it has affected me for a long time. I am doing the development and it is working on iOS, but in android I have a problem in[ this line](https://github.com/vallemar/NativeScript/commit/c4b3f4c2087416c121408bf510f8b8a09a98a3e1#diff-44160f1e0e1e7daf5bdbb721e722774aeeda6f45dc59c382869ceed266b84c76R173), it is always 0. I understand that the measurement is not available yet, but is there any way to recalculate when it is?

**Comment by farfromrefug** [comment]
@vallemar i think it is a tough one in the sense we dont have the logic to recompute a property after measurement of the parent. It would require quite some work i think.
Btw i have to say this can be avoided using width + margin. 
I never had the need for calc

**Comment by rigor789** [comment]
Yeah, this one seems a bit tricky to implement - what I imagine we would need is a way to pass a function as the property value that represents the `calc` values and can be scheduled after a layout. I think this specifically would require 2 layout passes, 1st where we lay everything out to get the measures and then another one to do the `calc`'s. 

What I imagine (and may not be necessary or the right way, just sharing my thoughts) could work is if we compiled the calc to a function (`new Function(...)`, where any `%` lengths are substituted with js code and converted to a valid js expression:

```js
// calc(100% - 10px);
// would become
const expressionFunction = new Function('view', 'property', `
  return (1 * view.parent.getActualSize()[property]) - (10)
  //      ^ percent converted to multiplier          ^ operator
  //        * parent property                           ^ literal
`)
```

So in the layout process, if we encounter a function value, we flag it to evaluate after an initial layout (and set it to 0 initially?)

The above Function can then be called with `expressionFunction(view, property)` and should return a number to use for the value...

--
This is all just ideas, how deep of a change this requires in core, I'm not sure.

**Comment by vallemar** [comment]
@rigor789 ok, I'm going to try to do something generic to also be able to adopt other situations in which the css has to be recalculated.
@farfromrefug I understand that it can be done in more ways, but when you come from the front it is useful to have all this because it is what you are used to.

I'm going to give it a spin anyway.

**Comment by rigor789** [comment]
@vallemar I think there needs to be special flags that are set when we come across a calc property that can't be evaluated ahead of time (ie with percentages) - then after a requestLayout pass, we need to trigger a new one probably in `layoutChanged` or maybe a lower level somewhere. I'm not sure where I would put it though, but the pseudo logic would be like:

- come across a calc that can't be evaluated aot
- convert/compile it to a function (mainly to prevent having to do it on every layout pass, only when the expression changes would a re-compile take place)
- set a flag like "REQUIRES_SECONDARY_LAYOUT_PASS" (name is debatable, depends on where it ends up being set/called, another name may be more appropiate)
- requestLayout runs, and if the flag is present, calc functions will default to 0 (if that makes sense?)
- requestLayout finishes, we check for the flag, if present remove it (probably would use bitwise flags) and then call requestLayout again
- this time the flag is no longer present, and we can call the calc function which has actual values to work with from the first layout pass now

**Comment by vallemar** [comment]
@rigor789  Ok, I think I understand what it says. I'm going to try to pull the string and do what it says. For now I will call it: REQUIRES_SECONDARY_LAYOUT_PASS, when I have a PR we will see more details. Thank you!
